File: tests/test-minecraft/src/main/java/pize/tests/minecraft/game/gui/components/ListView.java
package pize.tests.minecraft.game.gui.components;

import pize.Pize;
import pize.graphics.util.batch.TextureBatch;
import pize.math.Maths;

public class ListView extends MComponent{
    
    private float scrollX, scrollY;
    
    @Override
    protected void render(TextureBatch batch, float x, float y, float width, float height){
        batch.scissor.begin(228, x, y, width, height);
        
        scrollX = Maths.clamp(scrollX + Pize.mouse().getScrollX(Pize.keyboard()) * 50, -1000, 1000);
        scrollY = Maths.clamp(scrollY + Pize.mouse().getScrollY(Pize.keyboard()) * 50, -200, 0);
        setChildShift(scrollX, -scrollY);
    }
    
    @Override
    protected void renderEnd(TextureBatch batch){
        batch.scissor.end(228);
    }
    
    
    public boolean isHoverIgnoreChildren(){
        float mouseX = Pize.getX();
        float mouseY = Pize.getY();
        return !(mouseX < x || mouseY < y || mouseX > x + width || mouseY > y + height);
    }
    
}

File: tests/test-minecraft/src/main/kotlin/pize/tests/minecraft/game/gui/components/ListView.kt
package pize.tests.minecraft.game.gui.components

import pize.Pize
import pize.Pize.keyboard
import pize.Pize.mouse
import pize.graphics.util.batch.TextureBatch
import pize.math.Maths.clamp

class ListView : MComponent() {
    private var scrollX = 0f
    private var scrollY = 0f
    override fun render(batch: TextureBatch, x: Float, y: Float, width: Float, height: Float) {
        batch.scissor.begin(228, x.toDouble(), y.toDouble(), width.toDouble(), height.toDouble())
        scrollX = clamp(scrollX + mouse()!!.getScrollX(keyboard()!!) * 50, -1000f, 1000f)
        scrollY = clamp(scrollY + mouse()!!.getScrollY(keyboard()!!) * 50, -200f, 0f)
        setChildShift(scrollX, -scrollY)
    }

    override fun renderEnd(batch: TextureBatch) {
        batch.scissor.end(228)
    }

    val isHoverIgnoreChildren: Boolean
        get() {
            val mouseX = Pize.x.toFloat()
            val mouseY = Pize.y.toFloat()
            return !(mouseX < x || mouseY < y || mouseX > x + width || mouseY > y + height)
        }
}
